File: 2021/src/tasks/__tests__/day1.ts
import { realDataPath, test, testDataPath } from "../../util/test";
import { task1, task2 } from "../day1";

describe("task 1", () => {
  it("test data", test(task1, testDataPath("day1"), 7));
  it("real data", test(task1, realDataPath("day1"), 1266));
});

describe("task 2", () => {
  it("test data", test(task2, testDataPath("day1"), 5));
  it("real data", test(task2, realDataPath("day1"), 1217));
});

File: 2024/AoC2024/Sources/Tasks/Day1.swift
import Foundation
internal import Algorithms
import Framework

public final class Day1: Day<Void, Int> {
    public override func perform(data: String, task: Task, param: P) throws -> R {
        let (left, right) = try parse(data)
        
        return switch task {
        case .task1: task1(left: left, right: right)
        case .task2: task2(left: left, right: right)
        }
    }
    
    private func parse(_ data: String) throws -> ([Int], [Int]) {
        let numbers = try data.words.map(toInt)

        let left = numbers.striding(by: 2)
        let right = numbers.dropFirst().striding(by: 2)
        
        return (Array(left), Array(right))
    }
    
    private func task1(left: [Int], right: [Int]) -> Int {
        zip(left.sorted(), right.sorted())
            .map({ abs($0.0 - $0.1) })
            .reduce(0, +)
    }
    
    private func task2(left: [Int], right: [Int]) -> Int {
        left.map { leftValue in
            leftValue * right.count(where: { rightValue in leftValue == rightValue })
        }.reduce(0, +)
    }
}
